### PR TITLE
[batch-beta-dev] Implement TaskPerFile task factory.

### PR DESCRIPTION
**Pull request by xingwu1**

### CODE REVIEW

File: Documentation/BatchDocumentation/taskFactories.md [hunk 1]
```diff
@@ -173,4 +173,110 @@ Where the following tasks would be created:
...
+The task template into which the file URL/name will be substituted is defined in `repeatTask`. Substitutions are achieved
+through the use of placeholders. A placeholder for name substitutions is represented by `{keyword}`. The keyword here represents
+which part of file URL to be substituted. The supported keyword are:
+
+Keyword|Note|Example
+---|---|---
+`{url}`|The full URL of file location|http://account.blob.azure.com/container/path/blob.ext?sasToken
+`{filePath}`|The file name including the path (virtual directory)|path/blob.ext
+`{fileName}`|The file name only, without path|blob.ext
+`{fileNameWithoutExtension}`|The file name without last extension|blob
```
Comment (annatisch) [inline Documentation/BatchDocumentation/taskFactories.md]: Maybe put quote marks or something around the examples? Might not be necessary once it's rendered in table form...

File: lib/commands/batch/batch.templateUtils._js [hunk 2]
```diff
@@ -425,19 +412,133 @@ batchTemplateUtils.replacementParameter = function (data, parameters) {
     }
   });
 
+  return input;
+};
+
+/**
+ * Replace string with file value
+ * @param {string} data            The string to be replaced
```
Comment (annatisch) [inline lib/commands/batch/batch.templateUtils._js]: data => input

File: Documentation/BatchDocumentation/taskFactories.md [hunk 3]
```diff
@@ -173,4 +173,110 @@ Where the following tasks would be created:
...
+  }
+}
+```
+
+The list of files used to create the tasks are set in `source`. Similar to new `ResourceFiles` property, there are two ways to specify the file container in Azure Storage. 
+1. Specify the data stored at in linked storage under a file group can be simply referenced by file group name.
+2. Specify the full container URL include the SAS key which has to have List and Read permission. For example:
+```json
+"source": {
+	"container": "https://storage.blob.core.windows.net/container?sv=2015-04-05sig=tAp0r3I3SV5PbjpZ5CIjvuo1jdUs5xW"
```
Comment (annatisch) [inline Documentation/BatchDocumentation/taskFactories.md]: I know I should've raised this earlier - but I wonder if "container" should be "containerUrl" to be in keeping with the new ResourceFile schema. This was one of the changes recommended after review.

File: test/commands/batch/cli.batch.ncj-tests._js [hunk 4]
```diff
@@ -86,28 +71,58 @@ describe('cli', function () {
...
+      (function(){ templateUtils.replacementTransform(templateUtils.transformSweepString, "cmd {0}.mp3 {2}.mp3", [5, 10]); }).should.throw("The parameter pattern '{2}' is out of bound.");
+      (function(){ templateUtils.replacementTransform(templateUtils.transformSweepString, "cmd {}.mp3 {2}.mp3", [5, 10]); }).should.throw("The parameter pattern '{2}' is out of bound.");
+      (function(){ templateUtils.replacementTransform(templateUtils.transformSweepString, "cmd {{0}}}.mp3 {1}.mp3", [5, 10]); }).should.throw("Invalid use of bracket characters, did you forget to escape (using {{}})?");
+      (function(){ templateUtils.replacementTransform(templateUtils.transformSweepString, "cmd {0:3}.mp3 {1}.mp3", [-5, 10]); }).should.throw("The parameter '-5' is negative and cannot be used in pattern '{0:3}'.");
+      (function(){ templateUtils.replacementTransform(templateUtils.transformSweepString, "cmd {0:-3}.mp3 {1}.mp3", [5, 10]); }).should.throw("Invalid use of bracket characters, did you forget to escape (using {{}})?");      
+      done();
+    });
+
+    it('should correct replace file iternation command', function (done) {
```
Comment (annatisch) [inline test/commands/batch/cli.batch.ncj-tests._js]: 'should correctly replace file iteration command'

File: test/commands/batch/cli.batch.ncj-tests._js [hunk 5]
```diff
@@ -57,25 +57,10 @@ function getValueFromJson(obj, val) {
...
-    });
-    
-    afterEach(function (done) {
-      done();
-    });
 
     it('should expand template with parameter file', function (_) {
       this.interaction = new Interactor(this);
-      var templateFile = path.resolve(__dirname, '../../data/batch.job.simple.json');
+      var templateFile = path.resolve(__dirname, '../../../Documentation/BatchDocumentation/samples/ffmpeg/job.json');
```
Comment (annatisch) [inline test/commands/batch/cli.batch.ncj-tests._js]: Should we keep a version of this template in the data directory specifically for testing? We may eventually move the samples out to the samples repo.

File: Documentation/BatchDocumentation/samples/ffmpeg/job.perFile.json [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        "jobId": {
+            "type": "string",
+            "metadata": {
+                "description": "The name of Azure Batch job"
+            }
+        }
+    },
+    "job": {
+        "type": "Microsoft.Batch/batchAccounts/jobs",
+        "apiVersion": "2016-12-01",
```
Comment (xingwu1) [inline Documentation/BatchDocumentation/samples/ffmpeg/job.perFile.json]: All the template has this version though I know we don't use it at all.

Comment (annatisch) [inline Documentation/BatchDocumentation/samples/ffmpeg/job.perFile.json]: Oh really? I removed it from the Blender one but if the others still have that's fine :)

File: Documentation/BatchDocumentation/taskFactories.md (hunk 1 above)
Comment (xingwu1) [inline Documentation/BatchDocumentation/taskFactories.md]: it looks fine in table

File: Documentation/BatchDocumentation/taskFactories.md [hunk 7]
```diff
@@ -134,43 +134,149 @@ An example of a task factory with a two-dimensional sweep with zero-padding:
...
   }
-]
+```
+
+The list of files used to create the tasks are set in `source`. Similar to new `ResourceFiles` property, there are two ways to specify the file container in Azure Storage. 
+1. Specify the name of a file group to reference data stored in a linked storage account.
+2. Specify the full container URL include the SAS key which has to have List and Read permission. For example:
+```json
+  "source": {
+	  "container": "https://storage.blob.core.windows.net/container?sv=2015-04-05sig=tAp0r3I3SV5PbjpZ5CIjvuo1jdUs5xW"
```
Comment (annatisch) [inline Documentation/BatchDocumentation/taskFactories.md]: This also needs to be "containerUrl"

Comment (xingwu1) [inline Documentation/BatchDocumentation/taskFactories.md]: It is weird. It missed during the commit. I will update it.

**Pull request merged**